### ISSUE TITLE
Ansible substitutes YAML null value with "None" in templates

### ISSUE THREAD

**Issue by HouzuoGuo** [ansible/ansible#7984]
##### Summary:

In template task, Ansible substitutes YAML null value variables with text "None". This is not a sensible substitution.
##### Steps To Reproduce:
- Create a minimal file template `{{foo}}`
- Create a minimal ansible playbook, and assign variable foo to nothing `foo:`
- Run the playbook
##### Expected Results:

Ansible substitutes variable `foo` with empty string.
##### Actual Results:

Ansible substitutes variable `foo` with string "None".

Although "None" is meaningful in Python, but it is not a meaningful universal representation of null. A sensible choice is empty string.

**Comment by mpdehaan** [comment]
I would believe an empty string is much less clear, and Ansible is in fact keeping the internal representation as None.

I'm not entirely sure how you are looking at the output to come to the above conclusion, but either way, we would not be printing empty string for None as that would be unclear.

**Comment by HouzuoGuo** [comment]
Does any user actually rely on the behaviour which Ansible substitutes null variable with string "None"?

**Comment by srgvg** [comment]
I actually rely on it that is doesn't return the empty string if none.
None is an undefined value, the empty tring is defined.

**Comment by gaastonsr** [comment]
I agree. Empty string is a string.

None or null would make sense if `foo` was not defined as a variable.

**Comment by darkmoon03** [comment]
That would make sense if {%if foo is defined %} didn't evaluate to true.

If I've got 
foo: 
as a an (undefined) hostvar, and {%if foo is defined %} as a conditional in a template, it treats 'None' as a defined value.

NULL makes more sense, as does a empty string (which I would agree is defined though).

**Comment by bcoca** [comment]
if foo is not None is what you want, None is not undefined.

**Comment by darkmoon03** [comment]
No, not what I'd like.

Right now I set it to null (or don't set it all), but have to test vs 'None'.  Not even remotely intuitive.

It _is_ undefined.  But '{%if foo is defined %}' sees it as 'None' and treats it as defined, requiring a second check.

Oddly, this changed in the last month or two.  (I use this for interface config and haven't had to rebuild any interfaces recently.)

**Comment by bcoca** [comment]
This is part of YAML and Python, the parser will convert YAML null to Python None, Python does not equate None and Undefined, this is just how the language works, Ansible uses it for conditionals so it must follow the language rules.

**Comment by darkmoon03** [comment]
Why doesn't YAML convert a Null to a Python Undefined?  That would make much more sense.

How would you represent the text string "None" and differentiate it from from Undefined?

**Comment by sivel** [comment]
Undefined means that you literally haven't defined it anywhere.  By assigning it a value you have defined it.

`null` is equivalent to python `None`.  `None` does not mean it is undefined, it means it is defined as `None`.  To check if it is defined and truthy, one would have to use something like:

```
{% if foo is defined and foo %}
```

The string `"None"` evaluates as truthy and defined, the actual python `None` evaluates as "falsey" and defined.

**Comment by darkmoon03** [comment]
Ahh, that makes much more sense.  Thank you.

So, the conversion isn't picking up the actual value, but the display value.  I'd still call that a bug, since I didn't assign it a value any where, I just mentioned it.  But it's liveable as long as it's not truthy.  

Thanks again.   I'm off to rewrite some templates.

**Comment by einarc** [comment]
And remember kids, if you really want to be effective at Ansible, learn Python FIRST, especially if you come from the C and everything derived from it world where NULL=UNDEFINED (like a non instantiated object that has an existing class):

The key is here:

> This is part of YAML and Python, the parser will convert YAML null to Python None, Python does not equate None and Undefined, this is just how the language works, Ansible uses it for conditionals so it must follow the language rules.

And here:

>  Undefined means that you literally haven't defined it anywhere. By assigning it a value you have defined it.
> 
> null is equivalent to python None. None does not mean it is undefined, it means it is defined as None. To check if it is defined and truthy, one would have to use something like:
> 
> {% if foo is defined and foo %}
> **The string "None" evaluates as truthy and defined, the actual python None evaluates as "falsey" and defined.**

In other words:
 1.- If you've registered/set_fact a variable YOU'VE defined it, even if the YAML/JSON value is null or if the variable is a field of a bigger object.
2.-Do NOT compare against the "None" string, just evaluate the variable.
3.-If you're trying to evaluated a variable that you've registered it is ALREADY defined.

@bcoca @sivel  I think we need a BIG/ HUGE disclaimer on the Ansible Documentation telling people that if they want to use ansible for anything serious they should get their Python basics first. Goes without saying that it SHOULD be obvious but anyhow, you cannot find important details like this on either Ansible Docs or by Googling for Ansible methods because the information is at a deeper level: Python's.

In other news, can someone explain to me why this happens?

```
- name: restore rds instance from snapshot
  rds:
    command: restore
    snapshot: "{{ wp_db_snapshot }}"
    instance_name: "{{ wp_db_instance }}"
    region: "{{ region }}"
    subnet: "{{ rds_subnet }}"
    zone: "{{ rds_az }}"
    publicly_accessible: yes
    # wait: yes
    # wait_timeout: 8000
  when: not create_rds
  register: rds
  # async: 8000
  # poll: 15
  tags: rds_database

- name: Debug RDS facts
  rds:
    command: facts
    instance_name: '{{ wp_db_instance}}'
    region: "{{ region }}"
  register: rds_facts
  until: not(rds_facts.instance.endpoint is none)
  retries: 20
  delay: 30

- debug: var=rds_facts
```

Then it basically continues to iterate through rds:facts without changing the value of rds_facts.instance.endpoint even as the RDS instance goes through its states:
creating
backing-up
modifying
available

By the time the instance is on backing up the endpoint data is shown in the AWS console so how come the gather facts version of the rds module doesn't update the endpoint field UNTIL the instance is available and not when the endpoint is ACTUALLy available, even if the instance is not. This pretty much stops you from using an async approach where you can populate your app with the endpoint and create an ASG with it later on, which basically happens on the transition from creating to backing-up. I'm posting this as a bug on the appropriate place later on.

**Comment by gaastonsr** [comment]
I disagree, people should not need to learn Python to be effective with Ansible. It makes no sense.

**Comment by einarc** [comment]
@gaastonsr  Good luck figuring this out then. Unless you're saying the documentation should be covering this, Python and JINJA run ansible plain and simple fact. There's no way around that fact.

**Comment by einarc** [comment]
https://github.com/ansible/ansible-modules-core/issues/3865

**Comment by curaloucura** [comment]
@einarc I feel it's very unfair to claim that "people should learn python basics first".  We are talking more about specifics of some modules than "python basics" and even experienced python developers would run into trouble like we did.

I just ran into this problem that two other former developers implemented when declaring a database password.
In the global var files everything is a string, and it felt natural that leaving it empty we would get an empty string so nobody actually look at it until I was getting a weird bug and it took me a while to find it. Result, password and everything was being saved as "None" and when I actually had to check for an empty string, I found out the odd behaviour.

I know this ticket is closed but I may disagree that closing as a user misunderstanding of the internals is a good reason. My suggestion here is to put a mention for that in the Variable Docs just like the YAML gotcha. "If you want an empty string as value, use quotes instead of leaving the value empty, otherwise you will get a string 'None'.". Even if the variable files are not ini files, it feels like an ini file and anybody who works with ini files would expect that empty values are empty strings and here is counter-intuitive.
